shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
-        event_data::{
-            PoolRegistered as ContractPoolRegistered, TokensRegistered as ContractTokensRegistered,
-        },
-        Event as ContractEvent,
+    balancer_v2_factory::{
+        self, event_data::PoolCreated as ContractPoolCreated, Event as ContractEvent,
     },
-    BalancerV2Vault,
+    BalancerV2Vault, BalancerV2WeightedPool, BalancerV2WeightedPoolFactory,
 };
+use derivative::Derivative;
 use ethcontract::common::DeploymentInformation;
-use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
+use ethcontract::{
+    dyns::DynWeb3, Bytes, Event as EthContractEvent, EventMetadata, H160, H256, U256,
+};
+use mockall::*;
 use std::{
     collections::{HashMap, HashSet},
     fmt::Debug,
     ops::RangeInclusive,
 };
 use tokio::sync::Mutex;
 
-#[derive(Clone, Debug)]
-pub enum BalancerEvent {
-    PoolRegistered(PoolRegistered),
-    TokensRegistered(TokensRegistered),
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct PoolRegistered {
-    pub pool_id: H256,
+#[derive(Copy, Debug, Default, Clone, Eq, PartialEq, Hash)]
+pub struct PoolCreated {
     pub pool_address: H160,
-    pub specialization: PoolSpecialization,
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct TokensRegistered {
-    pub pool_id: H256,
-    pub tokens: Vec<H160>,
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 pub struct WeightedPool {
     pool_id: H256,
     pool_address: H160,
     tokens: Vec<H160>,
-    specialization: PoolSpecialization,
+    normalized_weights: Vec<U256>,
     block_created: u64,
 }
 
-#[derive(Clone, Debug, Default, PartialEq)]
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Hash)]
 pub struct WeightedPoolBuilder {
-    pool_registration: Option<PoolRegistered>,
-    tokens_registration: Option<TokensRegistered>,
-    /// Both Pool and Tokens Registered events are emitted in the same transaction
-    /// https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+    pool_creation: PoolCreated,
     /// block_number is only contained in the EventIndex
     block_created: u64,
 }
 
 impl WeightedPoolBuilder {
-    fn into_pool(self) -> Option<WeightedPool> {
-        if let (Some(pool_registration), Some(tokens_registration)) =
-            (self.pool_registration, self.tokens_registration)
-        {
-            return Some(WeightedPool {
-                pool_id: pool_registration.pool_id,
-                pool_address: pool_registration.pool_address,
-                tokens: tokens_registration.tokens,
-                specialization: pool_registration.specialization,
-                block_created: self.block_created,
-            });
-        }
-        None
+    async fn into_pool(self, data_fetcher: &dyn PoolDataFetching) -> Result<WeightedPool> {
+        let pool_address = self.pool_creation.pool_address;
+        let pool_data = data_fetcher.get_pool_data(pool_address).await?;
```

Comment:
Didn't we say this is something we want to do lazily when the pool is actually used not on every pool registration event (would add a roundtrip per pool on coldstart)

Reply:
Yes, you have mentioned this lazy approach in the past, but I am not convinced there is any value in storing a pool address without the known token information. Given that I think we need the token addresses immediately, I just thought we might as well fetch the rest of the pool info while we are at it.